Bounded Range Validation

apmoriarty opened the pull request:
Bounded range now verify that the lower bound sorts before the upper bound.

Because an invalid bounded range may be considered valid after normalization we cannot check bounded ranges immediately. A visitor was added to validate bounded ranges and that is run after terms are expanded from normalizers. 

- Added convenience methods to LiteralRange that compare the lower and upper bounds to test equivalency or error conditions
- UnmarkedBoundedRangeDetectionVisitor no longer operates on reference nodes per Jexl3 changes.
- Corrected many test queries and expected query plans that had improper bounded terms.
- Corrected a condition where a bounded range expands into a single term and would leave an improper node lineage (ref-and-singleterm)
- Added visitor to validate bounded range terms. It will throw an exception for an illegal range, but only warn about ranges with equivalent bounds that could be rewritten as equality terms.

While working this I noticed ExpandCompositeTerms was generating node trees with an invalid lineage, so opened https://github.com/NationalSecurityAgency/datawave/pull/2883 for that.

apmoriarty committed:
Added convenience methods to LiteralRange that compare the lower and
upper bounds to test equivalency or error conditions

UnmarkedBoundedRangeDetectionVisitor no longer operates on reference
nodes per Jexl3 changes.

Corrected many test queries and expected query plans that had improper
bounded terms.

Corrected a conditions where a bounded range expands into a single term
and would leave an improper node lineaage (ref-and-singleterm)

Added visitor to validate bounded range terms. It will throw an
exception for an illegal range, but only warn about ranges with
equivalent bounds that could be rewritten as equality terms.

jschmidt10 commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BoundedRangeIndexExpansionVisitor.java:
```diff
@@ -83,6 +83,16 @@ public Object visit(ASTAndNode node, Object data) {
             if (range != null) {
+
+                if (range.areBoundsEquivalent()) {
+                    log.warn("lower and upper bound are equivalent");
```
Can you add the range to this log as well?

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BaseIndexExpansionVisitor.java:
```diff
@@ -187,9 +190,37 @@ protected void rebuildFutureJexlNodes() {
+                if (!(newNode instanceof ASTOrNode || newNode instanceof ASTAndNode)) {
+                    if (isWrappedJunction(futureJexlNode.getOrigNode())) {
+                        JexlNode parent = futureJexlNode.jjtGetParent();
+                        JexlNode grandParent = parent.jjtGetParent();
```
could the grandparent be null?

apmoriarty replied:
It could, I'll add a check

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BaseIndexExpansionVisitor.java:
```diff
@@ -187,9 +190,37 @@ protected void rebuildFutureJexlNodes() {
             // if the parent is null, this is the root node, and we will handle that in the expand method
             if (futureJexlNode.jjtGetParent() != null) {
-                JexlNodes.replaceChild(futureJexlNode.jjtGetParent(), futureJexlNode, newNode);
+                if (!(newNode instanceof ASTOrNode || newNode instanceof ASTAndNode)) {
```
would love to see a bit more on why this is necessary and what you're correcting.

apmoriarty replied:
removed in next commit

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/UnmarkedBoundedRangeDetectionVisitor.java:
```diff
@@ -22,26 +34,30 @@ public static boolean findUnmarkedBoundedRanges(JexlNode script) {
+
+        // check for a bounded marker where the source node is not a range
+        Instance instance = QueryPropertyMarker.findInstance(node);
+        if (instance.isType(BOUNDED_RANGE)) {
+            LiteralRange<?> range = JexlASTHelper.findRange().getRange(node);
+            if (range == null) {
```
this doesn't look quite right. If we have a bounded range and the range is null set hasBounded = true?

apmoriarty replied:
This is equivalent to the old code. The LiteralRange.isRange() method returns true if the range is not equals to null. The variables are poorly named in this class, if you think a larger refactor would be better for clarity I'm happy to do that as part of this merge request. In essence, this visitor returns true if an unmarked bounded range was found.

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/UnmarkedBoundedRangeDetectionVisitor.java:
```diff
@@ -22,26 +34,30 @@ public static boolean findUnmarkedBoundedRanges(JexlNode script) {
+        // check for a range that is not marked
+        LiteralRange<?> range = JexlASTHelper.findRange().notDelayed().notMarked().getRange(node);
+        if (range != null && range.isBounded()) {
+            AtomicBoolean hasBounded = (AtomicBoolean) data;
```
if its not marked is there any concern about the data type cast?

apmoriarty replied:
No. But now I'm leaning towards a larger refactor for clarity.

FineAndDandy commented on warehouse/query-core/src/test/java/datawave/query/jexl/visitors/UnmarkedBoundedRangeDetectionVisitorTest.java:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * This illegal range is detected by the {@link ValidateBoundedRangeVisitor}
+     */
+    @Test
+    public void testBoundedMarkerWithInvalidRange() {
+        String query = "((_Bounded_ = true) && (FIELD >= '5' && FIELD <= '3'))";
+        test(false, query);
+    }
```
shouldn't this also throw an exception?

apmoriarty replied:
This visitor only checks to see if a bounded range is marked correctly. The BoundedRangeValidationVisitor will handle throwing an exception.

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/UnmarkedBoundedRangeDetectionVisitor.java:
```diff
@@ -22,26 +34,30 @@ public static boolean findUnmarkedBoundedRanges(JexlNode script) {
-        // determine if we have a marked range that is not actually a range
-        if (QueryPropertyMarker.findInstance(node).isType(BOUNDED_RANGE)) {
-            if (null != data && !JexlASTHelper.findRange().isRange(node)) {
+    public Object visit(ASTAndNode node, Object data) {
```
I'd like a couple of unit tests that hit all the code paths here.

apmoriarty replied:
Sure thing

FineAndDandy reviewed (changes requested):
a few questions and requests

apmoriarty committed:
Changed based on feedback:

log range when bounds are equivalent
refactor UnmarkedBoundedRangeDetectionVisitor for clarity
BaseIndexExpansionVisitor checks for null granparent nodes

apmoriarty commented:
Removing the integration tested label as the scope of changes grows

lbschanno reviewed (changes requested):
A couple comments and requests.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/validate/ValidateBoundedRangeVisitor.java:
```diff
@@ -0,0 +1,52 @@
+ * <p>
+ * This visitor must be run after bounded range expansion due to how normalization happens.
+ */
+public class ValidateBoundedRangeVisitor extends ShortCircuitBaseVisitor {
```
Will you be adding tests for this visitor?

apmoriarty replied:
added in next commit

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/jexl/LiteralRange.java:
```diff
@@ -4,17 +4,21 @@
  */
 public class LiteralRange<T extends Comparable<T>> implements Comparable<LiteralRange<T>> {
 
+    private static final Logger log = LoggerFactory.getLogger(LiteralRange.class);
```
As far as I can tell, this Logger instance isn't used anywhere.

apmoriarty replied:
removed in next commit

lbschanno reviewed (changes requested):
Review attempt 2, hopefully with the actual comments added this time.

apmoriarty committed:
Correct invalid node lineage in RegexIndexExpansionVisitor

Add unit tests for BoundedRange and Regex index expansion visitors

BaseIndexExpansionVisitor no longer inserts FutureJexlNodes into the
query tree

Add test for the ValidateBoundedRangeVisitor

RangeStream now considered bounded ranges for indexed fields to be
executable

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -982,6 +982,159 @@ private static Map<String,Object> getAssignments(JexlNode node, Map<String,Objec
+     * @param replacement
+     *            the replacement node
+     */
+    public static void replaceNodeSafely(JexlNode original, JexlNode replacement) {
```
Would you be willing to add tests for the new functions you added here?

apmoriarty replied:
Absolutely

lbschanno commented on warehouse/query-core/src/test/java/datawave/query/jexl/visitors/validate/ValidateBoundedRangeVisitorTest.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Test
+    public void testNoMarkerValidSource() {
```
`testNoMarkerValidSource()` and `testNoMakerInvalidSource()` are functionally equivalent since there is no marker in either of them, and I didn't see any logic in ValidateBoundedRangeVisitor that cares about EQ and LE/GE in non-marker nodes. I think these can be replaced with a single `testNoMarker()` test.

apmoriarty replied:
Can do

lbschanno reviewed (changes requested):
A couple extra comments. I like the test coverage!

apmoriarty committed:
Add and update tests per feedback

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/index/lookup/RangeStream.java:
```diff
@@ -917,9 +929,14 @@ private void populateCache() {
                 }
+
+                if (scannedKeys == 0) {
+                    log.error("no entries in num_shards cache");
```
should this be fatal?

apmoriarty replied:
sure

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -982,6 +982,161 @@ private static Map<String,Object> getAssignments(JexlNode node, Map<String,Objec
+        }
+
+        // if no 'legal' rewrite could occur with the parent, try the grandparent
+        JexlNode grandParent = parent.jjtGetParent();
+        if (grandParent != null) {
+            // reference expression, straight replacement
+            if (grandParent instanceof ASTReferenceExpression) {
+                JexlNodes.replaceChild(grandParent, parent, replacement);
+                return;
+            }
+
+            // similar junctions, merge children
+            if (isEquivalentJunctions(grandParent, replacement)) {
+                mergeJunctions(grandParent, parent, replacement);
+                return;
+            }
+        }
```
does this need to consider the type of node of the parent since the method requires any JexlNode?

apmoriarty replied:
No, it does not. The parent method can be used to replace reference expressions, and nodes, or nodes, or any leaf node. The check you see here simply ensure that any replacement produces a logically correct query tree, i.e., no adjacent junctions or reference expressions.

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -982,6 +982,161 @@ private static Map<String,Object> getAssignments(JexlNode node, Map<String,Objec
+            return;
+        }
+
+        // otherwise we have a reference expression and we need to go up a layer
+        JexlNode grandParent = parent.jjtGetParent();
+
+        if (grandParent != null) {
+            // check for the negated single term case: !(term)
+            // it is also unlikely that two reference expressions would sit next to each other in the tree, but it's
+            // not impossible
+            if (!(grandParent instanceof ASTNotNode) && !(grandParent instanceof ASTReferenceExpression)) {
+                JexlNodes.replaceChild(grandParent, parent, replacement);
+                return;
+            }
+        }
```
probably also need to explicitly check the type of the parent here?

apmoriarty replied:
Going back to the calling method, the node being replaced (the parent) can be any node.

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -982,6 +982,161 @@ private static Map<String,Object> getAssignments(JexlNode node, Map<String,Objec
+     * @param b
+     *            second JexlNode
+     */
+    public static void mergeJunctions(JexlNode parent, JexlNode a, JexlNode b) {
```
This has some preconditions that need to be checked. No guarantees that parent is related to a or b

apmoriarty replied:
The relationship check is handled in the calling method, but I see you point. I'll change this to private to enforce the preconditions.

FineAndDandy replied:
this is a public static method you can't rely on the calling method to validate the internals

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java:
```diff
@@ -982,6 +982,161 @@ private static Map<String,Object> getAssignments(JexlNode node, Map<String,Objec
+     */
+    public static void mergeJunctions(JexlNode parent, JexlNode a, JexlNode b) {
+        List<JexlNode> nodes = new ArrayList<>();
+        for (JexlNode child : JexlNodes.getChildren(parent)) {
+            if (child == a) {
+                // preserves node order
+                nodes.addAll(List.of(JexlNodes.getChildren(b)));
+            } else {
+                nodes.add(child);
+            }
+        }
```
this looks like it could lose nodes if parent contains more than just a? Also in that case might add a's nodes multiple times.

apmoriarty replied:
Nope. The two places this method is called look like this:
```
mergeJunctions(parent, original, replacement)
mergeJunctions(grandparent, parent, replacement)
```
This method is merging two junctions together while preserving node order. You can see this in the tests added for the JexlASTHelper, two notable examples are below

```
# replacement node preserved and original node order preserved
original: A == '1' && X == 'x' && D == '4'
replacement: B == '2' || C == '3'
A == '1' && (B == '2' || C == '3') && D == '4'

# replacement node is NOT preserved, it's children are inserted in-line into the parent node's children, original node order preserved
A == '1' && X == 'x' && D == '4'
B == '2' && C == '3'
A == '1' && B == '2' && C == '3' && D == '4'
```

FineAndDandy commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BaseIndexExpansionVisitor.java:
```diff
@@ -182,14 +180,31 @@ protected void rebuildFutureJexlNodes() {
             rebuildFutureJexlNode(futureJexlNode);
 
-            JexlNode newNode = futureJexlNode.getRebuiltNode();
+            if (futureJexlNode.jjtGetParent() == null) {
```
a future jexl node can never be the root of the tree?

apmoriarty replied:
That is correct. They were previously embedded in the node tree but not anymore because that was dumb.

FineAndDandy commented on warehouse/query-core/src/test/java/datawave/query/jexl/JexlASTHelperTest.java:
can you add some non-happy path tests please?

apmoriarty replied:
done

FineAndDandy reviewed (changes requested):
few more things

apmoriarty committed:
Updates from feedback
